### PR TITLE
Add support for Tornado 16X SQ air conditioner (0x4E2A)

### PR DESCRIPTION
**Pull request by felipediel**
This PR is based on @Enosh's work: https://github.com/mjg59/python-broadlink/pull/430. I was helping with the review and the PR ended up getting lost in the midst of so many changes and updates. He was being very cooperative, so I think now is the time for me to do my part. I rebased and cleaned up the code a little bit to have it merged soon.

This is a work in progress.

## Todo
- [x] Make tornado a generic HVAC class
- [x] Better names
- [x] Clean up IntEnums
- [x] Clean up encoders
- [ ] Create a better interface for the state
- [x] Change _decode() return value to `payload[10:p_len]`
- [ ] Fix CRC: cannot return negative values
- [ ] Decode UNK0, UNK1 and UNK2.
- [ ] Clean up
- [ ] Tests

### CODE REVIEW

**Comment by felipediel** [comment]
Hi @Enosh. I rebased the code so that it doesn't get lost in time. I took the opportunity to make some improvements and clean up a little. Still work? I wish we could have this in v0.17.0, because I don't know how many chances we will have after that.

**Comment by Enosh** [comment]
Hey, been away for a while, this looks very neat.
I did some testing, and found out some packet length issues but other than that everything works well.
Also isolated the "ifeel" option.

Since I can't push to your branch I put the commit [here](https://github.com/Enosh/python-broadlink/commit/ed4bcbb867cf235234378a1debf845084d913760).

File: broadlink/climate.py [hunk 1]
```diff
@@ -235,3 +238,217 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
...
+    def __init__(self, *args, **kwargs):
+        device.__init__(self, *args, **kwargs)
+        self.type = "HVAC"
+
+    def _crc(self, data: bytes) -> int:
+        """Calculate CRC of a byte object."""
+        s = sum([v if i % 2 == 0 else v << 8 for i, v in enumerate(data)])
+        # trim the overflow and add it to smallest bit
+        s = (s & 0xffff) + (s >> 16)
+        return (0xffff - s)  # TODO: fix: we can't return negative values
```
Comment (Enosh) [inline broadlink/climate.py:300]: I might be misunderstanding what you mean.
Screen Shot 2021-05-02 at 14 23 06
But if you mean be negative like in the screenshot, I think it'd be okay just to check that the packet is smaller than 2^18B.

**Comment by danielchalmers** [comment]
I have AUX Mini Split Air Conditioners with the same type that also can be controlled by AC Freedom

```
>python broadlink_discovery
Discovering...
###########################################
Unknown
# broadlink_cli --type 0x4e2a --host 192.168.1.61 --mac c8f7____7d8b
Device file data (to be used with --device @filename in broadlink_cli) :
0x4e2a 192.168.1.61 c8f7____7d8b

###########################################
Unknown
# broadlink_cli --type 0x4e2a --host 192.168.1.29 --mac a043____4a0d
Device file data (to be used with --device @filename in broadlink_cli) :
0x4e2a 192.168.1.29 a043____4a0d
```

I'd be more than happy to help test this if you have the time to rebase it in the future. Would love to see support for these devices!

**Comment by manio** [comment]
Hi,
I own [Ande](https://www.myande.pl/) unit. I am currently using it via AC Freedom app. It is **0x4e2a**.
Every opensource software I was trying so far, to communicate with this device fails with authentication failure.

I was trying this PR, here is my testing snippset:
``` python
#!/usr/bin/python
import broadlink

devices = broadlink.discover(timeout=5)
print(devices)
devices[0].auth()
```
and as the result:
```
[<hvac: Licensed manufacturer HVAC (0x4e2a) at [REDACTED]:80 | [REDACTED] | 奥克斯空调 | Locked>]
Traceback (most recent call last):
  File "/usr/src/klima-testy/python-broadlink/./monitor.py", line 6, in <module>
    devices[0].auth()
  File "/usr/src/klima-testy/python-broadlink/broadlink/device.py", line 167, in auth
    check_error(response[0x22:0x24])
  File "/usr/src/klima-testy/python-broadlink/broadlink/exceptions.py", line 137, in check_error
    raise exception(error_code)
broadlink.exceptions.AuthenticationError: [Errno -1] Authentication failed
```
Do you have any clue why is that?

**Comment by thewh1teagle** [comment]
Any update on this? does the library supports tornado ACs?

File: broadlink/climate.py (hunk 1 above)
Comment (felipediel) [reply broadlink/climate.py:300]: Yes, overflow is unlikely as we are not dealing with packets of this size. `& 0xFFFF` should be enough as a safeguard:

```python
     return (0xffff - s) & 0xffff
```

Comment (felipediel) [reply broadlink/climate.py]: I just found the polynomial, now we can do:
```python3
crc = CRC16.calculate(payload[0x02:p_len], polynomial=0x9BE4)
```

**Comment by felipediel** [comment]
Great, I believe we're all set now. Thank you very much, @enosh, and I apologize for the delay. You did a great job!

**Pull request merged**